.github/workflows/pr-gate.yml

```diff
@@ -0,0 +1,208 @@
+############### STATIC-ANALYSIS ###############
+###############################################
+  static-analysis:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Python 3.6
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.6
+    
+    - name: Install dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+
+    - name: Run flake8
+      # TODO: re-enable this flake8 blocks (turned off to get a draft of the pipeline infrastracture)
+      continue-on-error: true
+      run: |
+       tox -e flake8
+
+###############################################
+################## CPU-BUILD ##################
+###############################################
+  build-cpu:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        python: [3.6]
+        # different kind of tests are located in tests/<unit|integration|smoke> folders
+        test-kind: ['unit'] 
+        # pytest markers configured in tox.ini. See https://docs.pytest.org/en/6.2.x/example/markers.html
+        test-marker: ['not gpu and not spark and not notebooks', 'not gpu and notebooks and not spark']
+
+    steps:
+    - uses: actions/checkout@v2
+    ################# Run Python tests #################
+    - name: Use Python ${{ matrix.python }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python }}
+    
+    - name: Install build dependencies (tox)
+      run: |
+        python -m pip install --upgrade pip setuptools wheel
+        pip install tox
+    
+    - name: Run ${{ matrix.test-kind }} tests ('${{ matrix.test-marker }}')
+      # '-e py' will use the default 'python' executable found in system path
+      # for why using tox, see: https://tox.readthedocs.io/en/latest/index.html
+      # tox will do:
+      #   1. build and install source distribution (sdist)
+      #   2. run static analysis on the code (not implemented yet)
+      #   3. run all of the specified test environment (i.e. run tests in different pyversions, etc)
+      #   4. show test reports
+      run: |
+        tox -e py -- tests/${{ matrix.test-kind }} -m '${{ matrix.test-marker }}'
+      
+    - name: Prepare Code Coverage Report
+      run: |
+          mv .coverage '.coverage_${{ matrix.test-marker }}_${{ matrix.test-kind }}_${{ matrix.os }}_${{ matrix.python }}'
+          ls .coverage*
+      
+    - name: Upload Code Coverage
+      uses: actions/upload-artifact@v2
+      with:
+        name: code-cov
+        path: .coverage*
+
+###############################################
+################# SPARK-BUILD #################
+###############################################
+  build-spark:
+    runs-on: ${{ matrix.os }}
+    needs: static-analysis
+    strategy:
+      matrix:
+        os: [ubuntu-latest]
+        java: [8]
+        spark: [2.4.8]
```

Comment:
how do you link specific matrix combinations as being valid / invalid,
e.g. java=8 + spark=2.4.8 is good, but java=11 + spark.2.4.8 is bad

Reply:
Scott that is a good question! We can specify subset of combinations we want to remove from the matrix with the `exclude` keyword. Taking your example,
```yaml
java: [8, 11]
spark: [2.4.8, 3.0]
exclude:
   - java: 11
     spark: 2.4.8 
```
This will avoid the troublesome java11+spark2 combo

Reply:
ok, sounds reasonable